tests/test_size.py

```diff
@@ -1,170 +1,91 @@
-        self.assertSize(self.size, (1, 10, 0.1))
+    def change(self, dimension, value):
+        return self._replace(**{dimension: value})
 
-        # Layout has NOT been dirtied.
-        self.layout.dirty.assert_not_called()
 
-        # Set the height to something new
-        self.size.height = 20
-        self.assertSize(self.size, (1, 20, 0.1))
+BASE_SIZE = Size(width=1, height=2, ratio=0.1)
 
-        # Layout has been dirtied.
-        self.layout.dirty.assert_called_once_with(intrinsic_height=20)
 
-    def test_set_min_width(self):
-        self.size.width = at_least(10)
-        self.assertSize(self.size, (at_least(10), 2, 0.1))
+class TestBox:
+    pass
 
-        # Layout has been dirtied.
-        self.layout.dirty.assert_called_once_with(intrinsic_width=at_least(10))
 
-        # Clean the layout
-        self.layout.dirty.reset_mock()
+@pytest.fixture
+def box():
+    box = TestBox()
 
-        # Set the exact_width to the same value
-        self.size.width = at_least(10)
-        self.assertSize(self.size, (at_least(10), 2, 0.1))
+    box.maxDiff = None
 
-        # Layout has NOT been dirtied.
-        self.layout.dirty.assert_not_called()
+    box.layout = Mock()
+    box.size = BaseIntrinsicSize(layout=box.layout)
+    box.size._width, box.size._height, box.size._ratio = BASE_SIZE
 
-        # Set the exact_width to the same value, but not as a minimum
-        self.size.width = 10
-        self.assertSize(self.size, (10, 2, 0.1))
+    assert_size(box.size, BASE_SIZE)
 
-        # Layout has been dirtied.
-        self.layout.dirty.assert_called_once_with(intrinsic_width=10)
+    return box
 
-        # Clean the layout
-        self.layout.dirty.reset_mock()
 
-        # Set the exact_width to something new
-        self.size.width = at_least(20)
-        self.assertSize(self.size, (at_least(20), 2, 0.1))
+def assert_size(size, values):
+    assert (size.width, size.height, size.ratio) == values
 
-        # Layout has been dirtied.
-        self.layout.dirty.assert_called_once_with(intrinsic_width=at_least(20))
 
-    def test_set_min_height(self):
-        self.size.height = at_least(10)
-        self.assertSize(self.size, (1, at_least(10), 0.1))
+def test_at_least_repr():
+    assert repr(at_least(10)) == "at least 10"
 
-        # Layout has been dirtied.
-        self.layout.dirty.assert_called_once_with(intrinsic_height=at_least(10))
 
-        # Clean the layout
-        self.layout.dirty.reset_mock()
+def test_size_repr(box):
+    assert repr(box.size) == "(1, 2)"
+    box.size.width = at_least(10)
+    assert repr(box.size) == "(at least 10, 2)"
 
-        # Set the exact height to the same value
-        self.size.height = at_least(10)
-        self.assertSize(self.size, (1, at_least(10), 0.1))
 
-        # Layout has NOT been dirtied.
-        self.layout.dirty.assert_not_called()
+@pytest.mark.parametrize("dimension", ["width", "height"])
+@pytest.mark.parametrize("minimum", [True, False])
```

Comment:
I think this might parameterise better if you pass in the literal value for the dimension, (i.e., `10` or `at_least(10)`), rather than a boolean that adapts. I'm pretty sure the test works out the same, and you avoid the need for "if minimum" check in the test.

Reply:
Hm... I was mimicking the fact that the tests of minimum dimensions take an extra step that the absolute ones don't; they set the minimum size, then set the same-but-absolute size. I suppose performing that extra step all the time probably wouldn't hurt, but I'd need to pass in both `at_least(10)` and `10`, I think, in order to set it to both and check the differences.

Reply:
It's true that would be extra/duplicated work, but honestly I think it makes the test easier to understand. Having parameterised logic makes a test harder to parse, because you're not only trying to understand the logic you're testing - you're trying to understand the logic of the test itself. Having if statements in a test *at all* is a bit of a red flag - it's sometimes unavoidable, but if you *can* avoid it, you should IMHO.

Reply:
That said, maybe I am trying to be too smart with parametrization here. Would it be clearer to re-split the absolute and minimum size versions into two separate tests?

Reply:
I think it might be clearer to just split the absolute and minimum cases back into separate tests, honestly.

Reply:
Sure - that makes sense as well.